doc/development/adding_operators.rst

```diff
@@ -0,0 +1,356 @@
+                           do_queue=True, id=None):
+
+            # checking the inputs --------------
+
+            if do_flip and wire_flip is None:
+                raise ValueError("Expected a wire to flip; got None.")
+
+            # note: we use the framework-agnostic math library since
+            # trainable inputs could be tensors of different types
+            shape = qml.math.shape(angle)
+            if len(shape) > 1:
+                raise ValueError(f"Expected a scalar angle; got angle of shape {shape}.")
+
+            #------------------------------------
+
+            # do_flip is not trainable but influences the action of the operator,
+            # which is why we define it to be a hyperparameter
+            self._hyperparameters = {
+                "do_flip": do_flip
+            }
+
+            # we extract all wires that the operator acts on,
+            # relying on the Wire class arithmetic
+            all_wires = qml.wires.Wires(wire_rot) + qml.wires.Wires(wire_flip)
+
+            # The parent class expects all trainable parameters to be fed as positional
+            # arguments, and all wires acted on fed as a keyword argument.
+            # The id keyword argument allows users to give their instance a custom name.
+            # The do_queue keyword argument specifies whether or not
+            # the operator is queued when created in a tape context.
+            super().__init__(angle, wires=all_wires, do_queue=do_queue, id=id)
+
+        @property
+        def num_params(self):
+            # if it is known before creation, define the number of parameters to expect here,
+            # which makes sure an error is raised if the wrong number was passed
+            return 1
+
+        @staticmethod
+        def compute_decomposition(angle, wires, do_flip):  # pylint: disable=arguments-differ
+            # Overwriting this method defines the decomposition of the new gate, as it is
+            # called by Operator.decomposition().
+            # The general signature of this function is (*parameters, wires, **hyperparameters).
+            op_list = []
+            if do_flip:
+                op_list.append(qml.PauliX(wires=wires[1]))
+            op_list.append(qml.RX(angle, wires=wires[0]))
+            return op_list
+
+        def adjoint(self):
+            # the adjoint operator of this gate simply negates the angle
+            return FlipAndRotate(-self.parameters[0], self.wires[0], self.wires[1], do_flip=self.hyperparameters["do_flip"])
+
+The new gate can now be created as follows:
+
+.. code-block:: python
+
+    >>> op = FlipAndRotate(0.1, wire_rot="q3", wire_flip="q1", do_flip=True)
+    >>> op
+    FlipAndRotate(0.1, wires=['q3', 'q1'])
+    >>> op.decomposition()
+    [PauliX(wires=['q1']), RX(0.1, wires=['q3'])]
+    >>> op.adjoint()
+    FlipAndRotate(-0.1, wires=['q3', 'q1'])
+
+The new gate can be used in devices, which access the decomposition to interpret it:
```

Comment:
```suggestion
The new gate can be used with PennyLane devices. PennyLane checks with the device
whether it supports operations using the operation name.

- If the device registers support for an operation with the same name,
  PennyLane leaves the gate implementation up to the device. The device
  might have a hardcoded implementation, _or_ it may refer to one of the
  numerical representations of the operator (such as `operator.matrix`).
  
- If the device does not register support for an operation with the same
  name, PennyLane will automatically decompose the gate using `operator.decomposition`.
```

I think this is important to distinguish, since, for example, you cannot define your own `CNOT` gate as per the example above and overwrite the decomposition --- most PennyLane devices will simply keep using their own hardcoded CNOT implementation!

In fact, for users, the canonical way of implementing the above (if they are _only_ interested in a decomposition representation) would be:

```python
def FlipAndRotate(angle, wire_rot, wire_flip=None, do_flip=False):
    if do_flip and wire_flip is None:
        raise ValueError("Expected a wire to flip; got None.")

    shape = qml.math.shape(angle)
    if len(shape) > 1:
        raise ValueError(f"Expected a scalar angle; got angle of shape {shape}.")

    qml.PauliX(wires=wire_flip)
    qml.RX(angle, wires=wire_rot)

@qml.qnode(dev)
def circuit(angle):
    FlipAndRotate(angle, wire_rot="q1", wire_flip="q1")
    return qml.expval(qml.PauliZ("q1"))
```